### ISSUE TITLE
Short names according to RFC4512 section 6.2 not working as expected

### ISSUE THREAD

**Issue by fabianfreyer** [cannatag/ldap3#297]
According to [RFC4512 section 6.2](https://tools.ietf.org/html/rfc4512#section-6.2) I'd expect short names of attributes to point to the same property of an ldap3.abstract.entry.Entry. This is however not the case.

This becomes frustrating when dealing with schema that have the same attribute as a MUST, but are referring to it using different aliases. One example of this is `account` and `inetOrgPerson`:

```python
>>> obj_def = ObjectDef(['account', 'inetOrgPerson'], conn)
>>> obj_def
OBJ : account, inetOrgPerson [account (Structural) 0.9.2342.19200300.100.4.5, top (Abstract) 2.5.6.0, inetOrgPerson (Structural) 2.16.840.1.113730.3.2.2, organizationalPerson (Structural) 2.5.6.7, person (Structural) 2.5.6.6, top (Abstract) 2.5.6.0]
MUST: cn, objectClass, sn, userid
MAY : audio, businessCategory, carLicense, departmentNumber, description, destinationIndicator, displayName, employeeNumber, employeeType, facsimileTelephoneNumber, givenName, homePhone, homePostalAddress, host, initials, internationaliSDNNumber, jpegPhoto, l, labeledURI, localityName, mail, manager, mobile, o, organizationName, organizationalUnitName, ou, pager, photo, physicalDeliveryOfficeName, postOfficeBox, postalAddress, postalCode, preferredDeliveryMethod, preferredLanguage, registeredAddress, roomNumber, secretary, seeAlso, st, street, telephoneNumber, teletexTerminalIdentifier, telexNumber, title, uid, userCertificate, userPKCS12, userPassword, userSMIMECertificate, x121Address, x500uniqueIdentifier
```
From this, I'd expect if I specify `uid`, `sn` and `cn`, I'd not be missing mandatory attributes.

However:
```python
>>> writer = Writer(conn, obj_def)
>>> entry = writer.new("uid=foo,ou=users,dc=my-domain,dc=com")
>>> entry.sn="Name"
>>> entry.cn="Some Name"
>>> entry
DN: uid=foo,ou=users,dc=my-domain,dc=com - STATUS: Virtual, Missing mandatory attributes - READ TIME: <never>
    cn: <Virtual>
        CHANGES: [('MODIFY_REPLACE', ['Some Name'])]
    objectClass: <Virtual>
                 CHANGES: [('MODIFY_REPLACE', ['account', 'inetOrgPerson'])]
    sn: <Virtual>
        CHANGES: [('MODIFY_REPLACE', ['Name'])]
    uid: <Virtual>
         CHANGES: [('MODIFY_REPLACE', ['foo'])]
>>> writer.commit()
[...]
LDAPCursorError: mandatory attributes 'userid' missing in entry uid=foo,ou=users,dc=my-domain,dc=com
```

Trying to fix this by explicitely specifying `userid` does not work either:
```
>>> entry.userid="foo"
>>> writer.commit()
[...]
LDAPAttributeOrValueExistsResult: LDAPAttributeOrValueExistsResult - 20 - attributeOrValueExists - None - attribute 'uid' provided more than once - addResponse - None
```

This makes using the abstraction layer for this ~rather common combination~ of objectClasses basically unusable.
*EDIT*: The example is a bit contrived, since it's actually two structural objectclasses. However, the problem is also evident when creating e.g. an `account` entry with RDN `uid=foo` instead of `userid=foo`:
```python
>>> obj_def = ObjectDef(['account'], conn)
>>> writer = Writer(conn, obj_def)
>>> entry = writer.new("uid=foo,ou=users,dc=my-domain,dc=com")
[...]
LDAPCursorError: rdn type 'uid' not in objectclass definition
```

**Comment by fabianfreyer** [comment]
unfortunately I haven't had time to test this yet, but I could imagine a solution could be along the following lines:
```diff
diff --git a/ldap3/abstract/entry.py b/ldap3/abstract/entry.py
index 3961696..b1296c5 100644
--- a/ldap3/abstract/entry.py
+++ b/ldap3/abstract/entry.py
@@ -97,7 +97,8 @@ class EntryState(object):
         if self.status == STATUS_VIRTUAL or (self.status == STATUS_PENDING_CHANGES and self._initial_status == STATUS_VIRTUAL):  # checks if all mandatory attributes are present in new entries
             for attr in self.definition._attributes:
                 if self.definition._attributes[attr].mandatory:
-                    if (attr not in self.attributes or self.attributes[attr].virtual) and attr not in self.changes:
+                    if (attr not in self.attributes or self.attrbutes[attr].virtual) and attr not in self.changes \
+                        and (all([alias not in self.changes for alias in attr.oid_info.name]) if isinstance(attr.oid_info.name, list) else (attr not in self.changes)):
                         self.status = STATUS_MANDATORY_MISSING
                         break
```
Would this be the right way to go here? If yes I'd tidy this up and make it a pull request.

**Comment by fabianfreyer** [comment]
Alternatively, would it be a good idea to subclass `CaseInsensitiveDict` to be alias aware?

**Comment by cannatag** [comment]
I think the best solution is to subclass the CaseInsensitiveDict to something as CaseInsensitiveWithAliasDict so that sort names could be transparently managed. Will work on this in the weekend.

Thanks.
Giovanni

**Comment by fabianfreyer** [comment]
Awesome, thanks!

**Comment by cannatag** [comment]
please check code in dev. It should works for reader cursor. in a writer cursor it should works for modifying already existing attributes, not yet for creating new one. I still must check with aliases in rdn.

**Comment by cannatag** [comment]
Now it should fully work.

**Comment by cannatag** [comment]
released in v 2.2.1
